### ISSUE TITLE
incorrect bootstrap flags for using containerd

### ISSUE THREAD

**Issue by nbrational** [awsdocs/amazon-eks-user-guide#433]
Documentation mentions setting CloudFormation stack parameter BootstrapArguments to `--kubelet-extra-args --container-runtime containerd` during creation of unmanaged node group using steps [here](https://docs.aws.amazon.com/eks/latest/userguide/launch-workers.html). But when I try to do this kubelet fails to start:-

```
# journalctl -u kubelet --no-pager
-- Logs begin at Thu 2021-08-12 08:00:07 UTC, end at Thu 2021-08-12 08:05:19 UTC. --
Aug 12 08:01:01 ip-172-31-26-79.ec2.internal systemd[1]: Starting Kubernetes Kubelet...
Aug 12 08:01:01 ip-172-31-26-79.ec2.internal systemd[1]: Started Kubernetes Kubelet.
Aug 12 08:01:01 ip-172-31-26-79.ec2.internal kubelet[2857]: Usage:
Aug 12 08:01:01 ip-172-31-26-79.ec2.internal kubelet[2857]: kubelet [flags]
Aug 12 08:01:01 ip-172-31-26-79.ec2.internal kubelet[2857]: Flags:
Aug 12 08:01:01 ip-172-31-26-79.ec2.internal kubelet[2857]: --add-dir-header                                                                                            If true, adds the file directory to the header
```

I inspected the kubelet arguments and they looked incorrect (--container-runtime docker):-
```
# systemctl show --no-pager kubelet | grep Environment
Environment=KUBELET_ARGS=--node-ip=172.31.26.79 --pod-infra-container-image=602401143452.dkr.ecr.us-east-1.amazonaws.com/eks/pause:3.1-eksbuild.1 --v=2 KUBELET_EXTRA_ARGS=--container-runtime

# systemctl show --no-pager kubelet | grep ExecStart=
ExecStart={ path=/usr/bin/kubelet ; argv[]=/usr/bin/kubelet --cloud-provider aws --config /etc/kubernetes/kubelet/kubelet-config.json --kubeconfig /var/lib/kubelet/kubeconfig --container-runtime docker --network-plugin cni $KUBELET_ARGS $KUBELET_EXTRA_ARGS ; ignore_errors=no ; start_time=[Thu 2021-08-12 10:14:45 UTC] ; stop_time=[Thu 2021-08-12 10:14:45 UTC] ; pid=19619 ; code=exited ; status=255 }
```

Reviewed the code for [bootstrap.sh](https://github.com/awslabs/amazon-eks-ami/blob/master/files/bootstrap.sh) and understood that the script takes --container-runtime parameter as well as a --kubelet-extra-args parameter. Argument --container-runtime is used to set CONTAINER_RUNTIME variable, which is in-turn used to configure kubelet for either Docker or containerd runtime. As such, passing '--kubelet-extra-args --container-runtime containerd' to BootstrapArguments is incorrect and it must instead be set to '--container-runtime containerd'. Tested this by creating a new unmanaged node group using CloudFormation and this time the nodes could successfully join the cluster and containerd was used instead of Docker:-

```
# systemctl show --no-pager kubelet | grep Environment
Environment=KUBELET_ARGS=--node-ip=172.31.29.92 --pod-infra-container-image=602401143452.dkr.ecr.us-east-1.amazonaws.com/eks/pause:3.1-eksbuild.1 --v=2

# systemctl show --no-pager kubelet | grep ExecStart=
ExecStart={ path=/usr/bin/kubelet ; argv[]=/usr/bin/kubelet --cloud-provider aws --config /etc/kubernetes/kubelet/kubelet-config.json --kubeconfig /var/lib/kubelet/kubeconfig --container-runtime remote --container-runtime-endpoint unix:///run/dockershim.sock --network-plugin cni $KUBELET_ARGS $KUBELET_EXTRA_ARGS ; ignore_errors=no ; start_time=[Thu 2021-08-12 09:14:36 UTC] ; stop_time=[n/a] ; pid=2757 ; code=(null) ; status=0/0 }
```

As per [kubelet docs](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/), --container-runtime can be set to either docker or remote (for containerd).

Perhaps the documentation wishes to convey setting BootstrapArguments to `--container-runtime containerd --kubelet-extra-args <any kubelet args>` instead.

In summary, `BootstrapArguments: --container-runtime containerd` works. `BootstrapArguments: --kubelet-extra-args --container-runtime containerd` doesn't.

**Comment by pgasca** [comment]
Thank you @nbrational for the detailed analysis!

I looked through the page you linked as well as the entirety of current user documentation and I don't see any examples of both  `--kubelet-extra-args --container-runtime containerd` flags being used together anymore. I believe we made some changes that solved the issue after you submitted your request. Feel free to reopen if you disagree.
